### ISSUE TITLE
Cache misses when building vite library indirectly after building the same library directly (nx/devkit & chalk related) when using the in process hasher

### ISSUE THREAD

**Issue by gerarts** [nrwl/nx#18198]
### Current Behavior

1. Create a project with a vite app that imports a vite library
2. Nx build the vite library with daemon disabled (CI=true will do it)
3. Nx build the vite app with the daemon disabled

Results in a cache miss

### Expected Behavior

Having already built the library I would expect the app build to use the cached version

### GitHub Repo

https://github.com/gerarts/nx-cache-issue-minimal-example

### Steps to Reproduce

_Clone repo https://github.com/gerarts/nx-cache-issue-minimal-example and install deps_

```sh
# clear cache (if repeating)
rm -rf node_modules/.cache

# build lib (with CI flag enabled to force in-process-hash)
CI=true npx nx run-many -t build --projects=some-lib

# notice library gets built, observe hashed output in node_modules/.cache/nx

# build app (with CI flag enabled to force in-process-hash)
CI=true npx nx run-many -t build --projects=some-app

# notice library gets built again, even though output already exists for the direct build
```

### Nx Report

```shell
>  NX  Unable to locate swc-node or ts-node. Nx will be unable to run local ts files without transpiling.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node   : 18.12.1
   OS     : darwin-x64
   npm    : 8.19.2

   nx                 : 16.5.3
   @nx/js             : 16.5.3
   @nx/linter         : 16.5.3
   @nx/workspace      : 16.5.3
   @nx/devkit         : 16.5.3
   @nx/eslint-plugin  : 16.5.3
   @nx/react          : 16.5.3
   @nrwl/tao          : 16.5.3
   @nx/vite           : 16.5.3
   @nx/web            : 16.5.3
   typescript         : 5.1.6
   ---------------------------------------
   Community plugins:
   @mands/nx-playwright : 0.6.2
   ---------------------------------------
   Local workspace plugins:
   	 @cache-issue/some-lib
```

### Operating System

- [X] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I ran into different hashes when running through both modes, so I slightly modified `TaskHasherImpl` to collect the hashes generated by `hashExternalDependency` to create an object of the seen hashes.

When I diffed the resulting objects there seems to be a difference in:

```
"npm:@nx/vite->npm:@nx/devkit"
  "npm:@nx/devkit->npm:nx"
     "npm:nx->npm:chalk@4.1.2" // DIFF
```

on the direct library build, chalk seems to get imported, but when it runs through the app-build, this dependency is suddenly missing.

The same seems to be happening for

```
"npm:@vitest/utils->npm:pretty-format@29.6.1"
```

which in the same way only seems to be included when directly building the library.

This does not seem to happen when building through the daemon. I suspect it has something to do with the way one of the executors behaves when it thinks it is running in some non-interactive terminal mode (CI/CD style), but I haven't been able to drill down to the actual cause.

Linked the diff between the two `seen` objects, left being the in-library mode, and right being the through-app mode.

https://www.diffchecker.com/9d2YRCnV/

The two `projectGraph` objects passed are identical for both cases.

**Comment by gerarts** [comment]
Have updated the issue with a minimal example repo: https://github.com/gerarts/nx-cache-issue-minimal-example

**Comment by FrozenPandaz** [comment]
Fixed by #18020
